File: packages/expo-pico-notifications/android/src/main/java/expo/modules/pico/notifications/ExpoPicoNotificationsModule.kt
package expo.modules.pico.notifications

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.pico.BuildConfig

class ExpoPicoNotificationsModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoPicoNotifications")

        Constants(
            "notificationsSdkAvailable" to NotificationUtils.isNotificationSdkAvailable(),
            "notificationsSdkVersion"   to (NotificationUtils.getNotificationSdkVersion() ?: "unavailable")
        )

        Function("getPermissionStatus") {
            NotificationUtils.getPermissionStatus(appContext.reactContext)
        }

        AsyncFunction("requestPermissions") { promise: Promise ->
            NotificationsBridge.requestPermissions(
                onSuccess = { result -> promise.resolve(result) },
                onError   = { code, msg -> promise.reject(code, msg, null) }
            )
        }

        // IPPSPushClient.register(appId, fcmToken, callback)
        // Empty fcmToken is accepted by the SDK on PICO devices (the OS push
        // service issues a token without external FCM integration).
        AsyncFunction("registerForPushNotifications") { promise: Promise ->
            NotificationsBridge.registerForPushNotifications(
                appId = BuildConfig.PICO_APP_ID,
                fcmToken = "",
                onSuccess = { map -> promise.resolve(map) },
                onError   = { code, msg -> promise.reject(code, msg, null) },
            )
        }
    }
}

File: packages/expo-pico-notifications/src/ExpoPicoNotificationsModule.ts
import { NativeModule } from 'expo';
import { resolveNativeModule } from '@expo-pico/platform-service-common';

declare class ExpoPicoNotificationsModule extends NativeModule {
  readonly notificationsSdkAvailable: boolean;
  readonly notificationsSdkVersion: string;

  getPermissionStatus(): string;
  requestPermissions(): Promise<Record<string, unknown>>;
}

const { available, nativeModule } = resolveNativeModule<ExpoPicoNotificationsModule>('ExpoPicoNotifications');

export const NativeNotifications = available ? nativeModule : null;
export const notificationsNativeAvailable = available;
